From 9e894e1ff3e7e7c1d8298b6bc2959a701c96cf2d Mon Sep 17 00:00:00 2001
From: Stefan Assmann <sassmann@redhat.com>
Date: Tue, 2 Sep 2014 13:12:05 -0400
Subject: [ethernet] i40e/i40evf: fix rx descriptor status

Message-id: <1409663591-32531-83-git-send-email-sassmann@redhat.com>
Patchwork-id: 90559
O-Subject: [RHEL7.1 PATCH 082/148] i40e/i40evf: fix rx descriptor status
Bugzilla: 1091127
RH-Acked-by: Prarit Bhargava <prarit@redhat.com>
RH-Acked-by: Tony Camuso <tcamuso@redhat.com>

From: Jesse Brandeburg <jesse.brandeburg@intel.com>

As reported by Eric Dumazet, the driver is not masking the right
bits in the receive descriptor before it starts checking them.

This patch extends the mask to allow for the right bits to be
checked, and fixes the issue permanently via a define.

CC: Eric Dumazet <eric.dumazet@gmail.com>
Change-ID: I3274f7619057a950f468143e6d7e11b129f54655
Signed-off-by: Jesse Brandeburg <jesse.brandeburg@intel.com>
Signed-off-by: Jeff Kirsher <jeffrey.t.kirsher@intel.com>
Signed-off-by: Stefan Assmann <sassmann@redhat.com>
(cherry picked from commit c2451d7f55d667350bf6eeda5749ce9a239a9b3f)
---

Signed-off-by: Jarod Wilson <jarod@redhat.com>

diff --git a/drivers/net/ethernet/intel/i40evf/i40e_type.h b/drivers/net/ethernet/intel/i40evf/i40e_type.h
index 4fc9835..9c901fd 100644
--- a/drivers/net/ethernet/intel/i40evf/i40e_type.h
+++ b/drivers/net/ethernet/intel/i40evf/i40e_type.h
@@ -492,9 +492,6 @@ union i40e_32byte_rx_desc {
  } wb;  /* writeback */
 };
 
-#define I40E_RXD_QW1_STATUS_SHIFT 0
-#define I40E_RXD_QW1_STATUS_MASK (0x7FFFUL << I40E_RXD_QW1_STATUS_SHIFT)
-
 enum i40e_rx_desc_status_bits {
  /* Note: These are predefined bit offsets */
  I40E_RX_DESC_STATUS_DD_SHIFT  = 0,
@@ -511,9 +508,14 @@ enum i40e_rx_desc_status_bits {
  I40E_RX_DESC_STATUS_LPBK_SHIFT  = 14,
  I40E_RX_DESC_STATUS_IPV6EXADD_SHIFT = 15,
  I40E_RX_DESC_STATUS_RESERVED_SHIFT = 16, /* 2 BITS */
- I40E_RX_DESC_STATUS_UDP_0_SHIFT  = 18
+ I40E_RX_DESC_STATUS_UDP_0_SHIFT  = 18,
+ I40E_RX_DESC_STATUS_LAST /* this entry must be last!!! */
 };
 
+#define I40E_RXD_QW1_STATUS_SHIFT 0
+#define I40E_RXD_QW1_STATUS_MASK (((1 << I40E_RX_DESC_STATUS_LAST) - 1) \
+      << I40E_RXD_QW1_STATUS_SHIFT)
+
 #define I40E_RXD_QW1_STATUS_TSYNINDX_SHIFT   I40E_RX_DESC_STATUS_TSYNINDX_SHIFT
 #define I40E_RXD_QW1_STATUS_TSYNINDX_MASK (0x3UL << \
           I40E_RXD_QW1_STATUS_TSYNINDX_SHIFT)
-- 
1.7.1